working/augmentation-libraries/feature-specification.md

```diff
@@ -321,82 +322,150 @@ doesn't, it must repeat the type from the augmented definition__.
+        The term <code>'augmented' \<argumentList\></code> clause parses
+        as a `<primary>` expression.
+        When evaluated, it invokes the augmented constructor on the same
+        class and evaluates to the returned value.
+        Type inference is performed as for invoking a constructor with
+        the same function type.
+        The term `augmented` must not occur in optional parameter
+        default value expressions.
+    *   **Redirecting generative constructor**:
+        The term `augmented` has no meaning, and must not occur in
+        optional parameter default value expressions or in
+        target constructor arguments expressions.
+    *   **Redirecting factory constructor**: Contains no expressions.
+
+    The term <code>'augmented' \<argumentPart\></code> words parses
+    as a `<primary>` expression.
 
 In all relevant cases, if the augmented member is an instance member, it is
-invoked with the same value for `this`.
-
-Assume that the identifier `augmented` occurs in a source location where no
-enclosing declaration is augmenting. In this case, the identifier is taken
-to be a reference to a declaration which is in scope.
-
-*In other words, `augmented` is just a normal identifier when it occurs
-anywhere other than inside an augmenting declaration.*
-
-*Note that, for example, `augmented()` is an invocation of the augmented
-function or method when it occurs in an augmenting function or method
-declaration. (In the latter case, the augmenting method declaration must
-occur inside an augmenting type-introducing declaration, e.g., an
-augmenting class or mixin declaration). This is also true if `augmented()`
-occurs inside a local function declaration inside the body of that function
-or method declaration. We could say that `augmented` is a contextual
-reserved word because it is unable to refer to a declaration in scope when
-it occurs inside an augmenting declaration, it always has the special
-meaning which is associated with augmentations.*
-
-A compile-time error occurs if a declaration with the basename `augmented`
-occurs in a location where any enclosing declaration is augmenting. *This
-error is applicable to all such declarations, e.g., local functions, local
-variables, parameters, and type parameters.*
-
-Consider a non-augmenting member declaration _Dm_ that occurs inside an
-augmenting type declaration _Dt_. A compile-time error occurs if the
-identifier `augmented` occurs in _Dm_.
-
-*For example, inside `augment class C` we could have a declaration like
-`void f() {...augmented()...}`. This is an error because the outer
-`augment` forces the meaning of `augmented` to be about augmentation in the
-entire scope, but the method declaration is introductory and hence there is
-no earlier declaration to augment.*
+invoked with the same value for `this`. For a non-redirecting factory
+constructor, it's invoked with the same type parameter bindings for the
+surrounding type.
```

Comment:
It took me a while to realize why the type parameters are implicitly forwarded. I think it's because in the `augmented(...)` call, there isn't an obvious place to put them. Putting them after `augmented` would make it look like they are type parameters on the constructor itself, not the class. That would be a problem if/when we add support for generic constructors.

It does feel like there is a small expressiveness hole here. I can imagine someone wanting to augment a factory constructor and have it create an object with a slightly different (probably tighter) set of type arguments. It's probably fine to not support this.

Maybe add some non-normative text explaining the motivation here? (And if I didn't guess the motivation right, I would definitely like to read it. :) )

Reply:
Correct. Just as we are invoking instance members on the same `this`, we invoke the constructor on the same *class* (instantiated type).
That's definitely necessary for generative constructors, they are invariant in the type parameters of the object they initialize, if not by necessity then to avoid runtime checks in initialiers.

It also enforces that the `augmented` that is invoked is a part of *the same constructor*. It's not a different constructor you can call on a different type, it's a part of *this* constructor which is already being called on *this* type. It's a part that is *included* into this body by the invocation. (And inlining is a valid implementation approach.)

The same for instance members, the `augmented` has the same `this` because that is the `this` of this invocation, and the `augmented` is *part* of this invocation.
(We could allow `other.augmented(...)` if we really wanted to, as long as we know `other` extends this class - which is usually something we don't assume, so only if the classes are unimplementable, but then it looks more like overriding than augmenting.)

Reply:
I had a similar comment above... I hadn't considered that the type arguments are on the class and not the function... that does make sense. And in any case it is a very weird thing to return a class with a tighter generic type than was given, it probably shouldn't even be allowed.